### PR TITLE
fix(console,core,test): support filter by type for GET roles and apps APIs

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

support filter by type for GET roles and apps APIs and no longer reuse advanced search for both role and application filtering by type, made the following changes:
1. add optional `type` as GET /roles API's query parameter. Show all roles (including both m2m app roles and user roles if `type` is not specified)
2. add optional `types` as GET /applications API's query parameter (list all apps if not presented)
3. remove the monkey-patched `fieldsTypeMapping` input parameters when parsing query parameters for advanced search. Usually, the string values used for advanced search apply to multiple fields in DB records, and in most cases, the non-exact match mode is applied. While `type` is likely an enum type and needs an exact match mode, which means reusing advanced search for `type` filtering is not a good idea.

## Testing

Tested locally and CI.

Also added the following integration test cases:

Console:
1. search role on the roles listing page.
2. search apps on the roles details page, m2m app tab.
3. unassign m2m apps from a role on the role details page.
4. unassign a role from m2m apps on the app details page.

API (add search with keyword test cases):
1. GET /applications/:appId/roles
2. GET /users/:userId/roles
3. GET /roles/:roleId/applications
4. GET /roles/:roleId/users
ONLY add API test cases for using keywords (match the use case of the console "Search" component on corresponding pages).

## Checklist

- [ ] `.changeset`
- [ ] unit tests
- [x] integration tests
- [ ] necessary TSDoc comments

### CODE REVIEW

File: packages/console/src/components/RoleEntitiesTransfer/components/SourceEntitiesBox/index.tsx [hunk 1]
```diff
@@ -48,17 +48,10 @@ function SourceEntitiesBox<T extends User | Application>({
...
-      : buildUrl(`api/applications`, {
-          excludeRoleId: roleId,
-          ...commonSearchParams,
-          'search.type': ApplicationType.MachineToMachine,
-          'mode.type': 'exact',
-        })
+    buildUrl(`api/${roleType === RoleType.User ? 'users' : 'applications'}`, {
+      excludeRoleId: roleId,
+      ...commonSearchParams,
+    })
```
Comment (gao-sun) [inline packages/console/src/components/RoleEntitiesTransfer/components/SourceEntitiesBox/index.tsx:54]: don't need to pass `types=` for applications here?

File: packages/core/src/queries/application.ts [hunk 2]
```diff
@@ -48,14 +44,19 @@ const buildConditionArray = (conditions: SqlSqlToken[]) => {
 };
 
 export const createApplicationQueries = (pool: CommonQueryMethods) => {
-  const countApplications = async (search: Search, excludeApplicationIds: string[]) => {
+  const countApplications = async (
+    search: Search,
+    excludeApplicationIds: string[],
+    types?: ApplicationType[]
```
Comment (gao-sun) [inline packages/core/src/queries/application.ts:50]: better to add a comment block for describing each parameter and mention conditions are joined by `and`

File: packages/integration-tests/src/api/application.ts [hunk 3]
```diff
@@ -26,10 +26,7 @@ export const createApplication = async (
 
 export const getApplications = async (types?: ApplicationType[]) => {
   const searchParams = new URLSearchParams(
-    conditional(
-      types &&
-        types.length > 0 && [...types.map((type) => ['search.type', type]), ['mode.type', 'exact']]
-    )
+    conditional(types && types.length > 0 && types.map((type) => ['types', type]))
   );
```
Comment (gao-sun) [inline packages/integration-tests/src/api/application.ts:30]: i assume the issue was that application search with `type` hasn't been covered by test cases. it looks like we are not adding api test cases to cover this situation

File: packages/integration-tests/src/tests/console/machine-to-machine-rbac/utils.ts [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
   waitForToast,
 } from '#src/ui-helpers/index.js';
+import { expectNavigation, appendPathname } from '#src/utils.js';
 
 export const createM2mRoleAndAssignPermissions = async (
   page: Page,
   { roleName, roleDescription }: { roleName: string; roleDescription: string },
-  apiResources: Array<{ apiResourceName: string; permissionName: string }>
+  apiResources: Array<{ apiResourceName: string; permissionName: string }>,
+  backToListingPage = false // Will be on the details page of previously created role, need to set to `true` to go back to roles listing page.
```
Comment (gao-sun) [inline packages/integration-tests/src/tests/console/machine-to-machine-rbac/utils.ts:15]: you can use TSDoc to describe this parameter

File: packages/console/src/components/RoleEntitiesTransfer/components/SourceEntitiesBox/index.tsx (hunk 1 above)
Comment (darcyYe) [reply packages/console/src/components/RoleEntitiesTransfer/components/SourceEntitiesBox/index.tsx]: Updated, built the search parameter in the parent component.

File: packages/core/src/queries/application.ts (hunk 2 above)
Comment (darcyYe) [reply packages/core/src/queries/application.ts:58]: Updated.

File: packages/integration-tests/src/api/application.ts (hunk 3 above)
Comment (darcyYe) [reply packages/integration-tests/src/api/application.ts:30]: In this PR, we added console test cases to ensure the functionality of the search component.
Also add more API test cases for search assigned entities with keyword.

File: packages/core/src/queries/application.ts [hunk 5]
```diff
@@ -48,24 +44,48 @@ const buildConditionArray = (conditions: SqlSqlToken[]) => {
...
 
 export const createApplicationQueries = (pool: CommonQueryMethods) => {
-  const countApplications = async (search: Search, excludeApplicationIds: string[]) => {
+  /**
+   * Get the number of applications that match the search conditions, conditions are joined in `and` mode.
+   *
+   * @param search The search config object, can apply to `id`, `name` and `description` field for application
+   * @param excludeApplicationIds Exclude applications with these ids
+   * @param types Optional array of {@link ApplicationType}, filter applications by types, if not provided, all types will be included
+   * @returns The number of applications that match the search conditions
```
Comment (gao-sun) [inline packages/core/src/queries/application.ts:53]:
```suggestion
   * @param search The search config object, can apply to `id`, `name` and `description` field for application
   * @param excludeApplicationIds Exclude applications with these ids.
   * @param types Optional array of {@link ApplicationType}, filter applications by types, if not provided, all types will be included.
   * @returns A Promise that resolves the number of applications that match the search conditions.
```

File: packages/core/src/queries/application.ts [hunk 6]
```diff
@@ -48,24 +44,48 @@ const buildConditionArray = (conditions: SqlSqlToken[]) => {
...
 
+  /**
+   * Get the list of applications that match the search conditions, conditions are joined in `and` mode.
+   *
+   * @param search The search config object, can apply to `id`, `name` and `description` field for application
+   * @param excludeApplicationIds Exclude applications with these ids
+   * @param types Optional array of {@link ApplicationType}, filter applications by types, if not provided, all types will be included
+   * @param limit limit of the number of applications in each page
+   * @param offset offset of the applications in the result
+   * @returns The list of applications that match the search conditions
```
Comment (gao-sun) [inline packages/core/src/queries/application.ts:83]:
```suggestion
   * @param search The search config object, can apply to `id`, `name` and `description` field for application
   * @param excludeApplicationIds Exclude applications with these ids.
   * @param types Optional array of {@link ApplicationType}, filter applications by types, if not provided, all types will be included.
   * @param limit Limit of the number of applications in each page.
   * @param offset Offset of the applications in the result.
   * @returns A Promise that resolves the list of applications that match the search conditions.
```

File: packages/core/src/routes/application.ts [hunk 7]
```diff
@@ -69,14 +85,14 @@ export default function applicationRoutes<T extends AuthedRouter>(
       );
 
       if (paginationDisabled) {
-        ctx.body = await findApplications(search, excludeApplicationIds);
+        ctx.body = await findApplications(search, excludeApplicationIds, types);
```
Comment (gao-sun) [inline packages/core/src/routes/application.ts:88]: double check we have an issue to remove the `excludeApplicationIds` parameter (use excludeRoleId instead)

File: packages/integration-tests/src/api/role.ts [hunk 8]
```diff
@@ -58,8 +58,17 @@ export const assignScopesToRole = async (scopeIds: string[], roleId: string) =>
...
   authedAdminApi.delete(`roles/${roleId}/scopes/${scopeId}`);
 
-export const getRoleUsers = async (roleId: string) =>
-  authedAdminApi.get(`roles/${roleId}/users`).json<User[]>();
+/**
+ * Get users assigned to the role.
+ *
+ * @param roleId Concerned role id
+ * @param keyword Search among all users (on `id`, `name` and `description` fields) assigned to the role with `keyword`
+ * @returns All users which contains the keyword assigned to the role
```
Comment (gao-sun) [inline packages/integration-tests/src/api/role.ts:66]:
```suggestion
 * Get users assigned to the role.
 *
 * @param roleId Concerned role id.
 * @param keyword Search among all users (on `id`, `name` and `description` fields) assigned to the role with `keyword`.
 * @returns A Promise that resolves all users which contains the keyword assigned to the role.
```

File: packages/integration-tests/src/api/role.ts [hunk 9]
```diff
@@ -69,8 +78,17 @@ export const assignUsersToRole = async (userIds: string[], roleId: string) =>
...
 
-export const getRoleApplications = async (roleId: string) =>
-  authedAdminApi.get(`roles/${roleId}/applications`).json<Application[]>();
+/**
+ * Get apps assigned to the role.
+ *
+ * @param roleId Concerned role id
+ * @param keyword Search among all m2m apps (on `id`, `name` and `description` fields) assigned to the role with `keyword`
+ * @returns All m2m apps which contains the keyword assigned to the role
+ */
```
Comment (gao-sun) [inline packages/integration-tests/src/api/role.ts:87]:
```suggestion
 * @param roleId Concerned role id.
 * @param keyword Search among all m2m apps (on `id`, `name` and `description` fields) assigned to the role with `keyword`.
 * @returns A Promise that resolves all m2m apps which contains the keyword assigned to the role.
 */
```

File: packages/integration-tests/src/api/application.ts (hunk 3 above)
Comment (gao-sun) [reply packages/integration-tests/src/api/application.ts:30]: API tests and UI tests have no conflict though

**Review by gao-sun — COMMENTED** [review]
would you like to add a changeset to describe this fix?

File: packages/integration-tests/src/api/application.ts (hunk 3 above)
Comment (darcyYe) [reply packages/integration-tests/src/api/application.ts]: API tests added (test cases covered filtering the roles/applications by `type` and `search` parameters, which should be the same as the actual API calls triggering by the console when using the search component).

File: .changeset/sixty-ladybugs-exercise.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+---
+"@logto/integration-tests": patch
+"@logto/console": patch
+"@logto/core": patch
+---
+
+Bug fix:
+core/console: fix 500 error when using search component in console to filter both roles and applications.
```
Comment (gao-sun) [inline .changeset/sixty-ladybugs-exercise.md:8]:
```suggestion
fix 500 error when using search component in console to filter both roles and applications.
```

changeset itself indicates the changed packages i think

**Review by gao-sun — APPROVED** [review]

**Pull request merged**